File: model/src/main/java/net/contrapt/auction/model/BidderRepository.java
package net.contrapt.auction.model;

import org.springframework.data.jpa.domain.Specification;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;

import java.util.List;

/**
 * Created by msimmons on 10/9/14.
 */
public interface BidderRepository extends CrudRepository<Bidder, Long>, JpaSpecificationExecutor<Bidder> {

    public Bidder findByName(String name);

    public Bidder findByContactEmail(String email);

    @Query(
          "select NEW net.contrapt.auction.model.BidderSummary(bidder.id, bidder.name, sum(bid.amount), sum(payment.amount)) " +
          "from Bidder bidder FETCH ALL PROPERTIES " +
          "left outer join bidder.winningBids as bid "+
          "left outer join bidder.payments as payment group by bidder"
    )
    public List<BidderSummary> findAllSummary();
}

File: model/src/main/kotlin/net/contrapt/auction/model/BidderRepository.kt
package net.contrapt.auction.model

import org.springframework.data.jpa.repository.JpaSpecificationExecutor
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository

/**
 * Created by msimmons on 10/9/14.
 */
interface BidderRepository : CrudRepository<Bidder, Long>, JpaSpecificationExecutor<Bidder> {

    fun findByName(name: String): Bidder

    fun findByContactEmail(email: String): Bidder

    @Query("""select NEW net.contrapt.auction.model.BidderSummary(bidder.id, bidder.name, sum(bid.amount), sum(payment.amount))
from Bidder bidder FETCH ALL PROPERTIES
left outer join bidder.winningBids as bid
left outer join bidder.payments as payment group by bidder""")
    fun findAllSummary(): List<BidderSummary>
}
